include/flatbuffers/grpc.h

```diff
@@ -23,50 +23,271 @@
 #include "grpc++/support/byte_buffer.h"
 #include "grpc/byte_buffer_reader.h"
 
+namespace flatbuffers {
 namespace grpc {
 
+enum AddRef { ADD_REF };
+enum StealRef { STEAL_REF };
+
+// Message is a typed wrapper around a buffer that manages the underlying
+// `grpc_slice` and also provides flatbuffers-specific helpers such as `Verify`
+// and `GetRoot`. Since it is backed by a `grpc_slice`, the underlying buffer
+// is refcounted and ownership is be managed automatically.
+template <class T>
+class Message {
+ public:
+  inline Message() : slice_(grpc_empty_slice()) {}
+
+  inline Message(grpc_slice slice, AddRef) : slice_(grpc_slice_ref(slice)) {}
+
+  inline Message(grpc_slice slice, StealRef) : slice_(slice) {}
+
+  inline Message(const Message &other) : slice_(grpc_slice_ref(other.slice_)) {}
+
+  inline Message(Message &&other) : slice_(other.slice_) {
+    other.slice_ = grpc_empty_slice();
+  }
+
+  inline Message &operator=(const Message &other) {
+    slice_ = grpc_slice_ref(other.slice_);
+    return *this;
+  }
+
+  inline Message &operator=(Message &&other) {
+    slice_ = other.slice_;
+    other.slice_ = grpc_empty_slice();
+    return *this;
+  }
+
+  inline ~Message() { grpc_slice_unref(slice_); }
+
+  inline const uint8_t *mutable_data() const {
+    return const_cast<uint8_t*>(data());
+  }
+
+  inline const uint8_t *data() const {
+    if (slice_.refcount != nullptr) {
```

Comment:
Unfortunately the structs in the union have different layouts (i.e. length is at different places). I'm guessing they did this deliberately either for optimization or to prevent people from making the assumption that they can rely on a consistent layout.

Or, maybe I'm misunderstanding your suggestion?